File: mrcp/led.py
from mrcp.panel import *
from mrcp.points import *
from mrcp.config import *

class Led(BaseElement):
    def __init__(self, pos=Point(0, 0), color=LED_COLOR) -> None:
        super().__init__(pos=pos, color=color)

    def paint(self):

        found = searchLed(self._panel, self._pos)

        #print("Found", found, self)
        if found != self:
            # there is another wich is not self
            # only the first will be painted
            return

        pos = self._pos.toCoords()
        dwg = self._panel._dwg
        cutLayer = self._panel._cLayer
        ledLayer = self._panel._lLayer
        color = self._color
        circle = dwg.circle(center=pos, r=LED_SIZE/2, fill=color)
        ledLayer.add(circle)
        circle = dwg.circle(center=pos, r=LED_SIZE/2, stroke=COLOR_CUT,
                            stroke_width=0.2, fill="none")
        cutLayer.add(circle)
        circle = dwg.circle(center=pos, r=LED_SIZE/2+LED_MARGIN, stroke=COLOR_ENGRAVE,
                    stroke_width=0.2, fill="none")
        cutLayer.add(circle)

        return super().paint()

def searchLed(panel,pos=Point(0,0)) -> Led:
    pos = pos.toCoords()
    for obj in panel._paintable:
        if isinstance(obj,Led):
            if pos == obj._pos.toCoords():
                return obj
    return None

File: mrcp-painter-tco/src/Led.ts
import { PainterTco, TcoPainterHint } from "./PainterTco";
import { tcoPoint } from "./Point";
import { t_TcoBasicHint } from "./Types";

export type t_LedHint=t_TcoBasicHint

const defaultLedHint:t_LedHint={
    pos:undefined,
    type:undefined,
    color:undefined
}
export class TcoLedHint extends TcoPainterHint<t_LedHint>{
    constructor(hint:t_LedHint){
        super({...defaultLedHint, ...hint})
        
    }
    paintSelf(paper: PainterTco) {
        if(this.hint.color==undefined)
            this.hint.color=paper.options.ledColor;

        /* Evitar pintar dos veces (tiene que ver con el grabado laser)
        found = searchLed(self._config,self._panel, self._pos)

        if found != self:
            # there is another wich is not self
            # only the first will be painted
            return
        */
        let p_pos= tcoPoint(this.hint.pos)
        let [x,y] = p_pos.toCoords(paper.options)
        let dwg = paper.paper;
        let cutLayer = paper.layers.cut;
        let ledLayer = paper.layers.led;
        let trackLayer = paper.layers.track;
        let circle = dwg.circle(x,y, paper.options.ledSize/2).attr({stroke:"none",fill:this.hint.color})
        ledLayer.add(circle)

        circle = dwg.circle(x,y, paper.options.ledSize/2).attr({fill: "none",stroke: paper.options.cutColor,strokeWidth: 0.2}) 
        cutLayer.add(circle)

        circle = dwg.circle(x,y, paper.options.ledSize/2+paper.options.ledMargin).attr({fill: "none",stroke: paper.options.engraveColor,strokeWidth: 0.2});
        cutLayer.add(circle)

        circle = dwg.circle(x,y, paper.options.ledSize/2).attr({stroke:"white",fill:this.hint.color, strokeWidth: 0.2})
        trackLayer.add(circle)
    }

}